server/json-rpc/src/lib.rs

```diff
@@ -37,86 +51,221 @@ impl fmt::Display for SubgraphRemoveParams {
+        info!(self.logger, "Received subgraph_deploy request"; "params" => params.to_string());
+
+        if should_check_auth()
+            && Some(&auth.bearer_token) != self.authorized.read().unwrap().get(&params.name)
+        {
+            return Box::new(future::err(json_rpc_error(
+                JSON_RPC_UNAUTHORIZED_ERROR,
+                "Auth token is invalid".to_owned(),
+            )));
+        }
+
+        Box::new(
+            self.provider
+                .deploy(params.name, format!("/ipfs/{}", params.ipfs_hash))
+                .map_err(|e| json_rpc_error(JSON_RPC_DEPLOY_ERROR, e.to_string()))
+                .map(|_| Ok(Value::Null))
+                .flatten(),
+        )
     }
 
     /// Handler for the `subgraph_remove` endpoint.
     fn remove_handler(
+        &self,
         params: SubgraphRemoveParams,
-        provider: Arc<impl SubgraphProvider>,
-        logger: Logger,
-    ) -> impl Future<Item = Value, Error = jsonrpc_core::Error> {
-        info!(logger, "Received subgraph_remove request"; "params" => params.to_string());
-        provider
-            .remove(params.name_or_id)
-            .map_err(|e| json_rpc_error(1, e.to_string()))
-            .map(|_| Ok(Value::Null))
-            .flatten()
+        auth: AuthorizationHeader,
+    ) -> Box<Future<Item = Value, Error = jsonrpc_core::Error> + Send> {
+        info!(self.logger, "Received subgraph_remove request"; "params" => params.to_string());
+
+        if should_check_auth()
+            && Some(&auth.bearer_token) != self.authorized.read().unwrap().get(&params.name_or_id)
+        {
+            return Box::new(future::err(json_rpc_error(
+                JSON_RPC_UNAUTHORIZED_ERROR,
+                "Auth token is invalid".to_owned(),
+            )));
+        }
+
+        Box::new(
+            self.provider
+                .remove(params.name_or_id)
+                .map_err(|e| json_rpc_error(JSON_RPC_REMOVE_ERROR, e.to_string()))
+                .map(|_| Ok(Value::Null))
+                .flatten(),
+        )
+    }
+
+    /// Handler for the `subgraph_authorize` endpoint.
+    ///
+    /// Takens subgraph name and returns an auth token that can be used to
+    /// add/remove subgraphs under the input name.
+    ///
+    /// Requires bearer authorization with the master token.
+    fn authorize_handler(
+        &self,
+        params: SubgraphAuthorizeParams,
+        auth: AuthorizationHeader,
+    ) -> Result<Value, jsonrpc_core::Error> {
+        info!(self.logger, "Received subgraph_authorize request"; "params" => params.to_string());
+
+        let master_token = env::var(GRAPH_MASTER_TOKEN_VAR);
+        match &master_token {
+            Ok(master_token) if *master_token == auth.bearer_token => (), // Authorized.
+            Ok(_) => {
+                return Err(json_rpc_error(
+                    JSON_RPC_UNAUTHORIZED_ERROR,
+                    "auth token is invalid".to_owned(),
```

Comment:
I guess in this case "auth token" is the right term, although I'd make it `Authorization token is invalid`.